docs/rfc/0001-exporter-interface.md

```diff
@@ -0,0 +1,192 @@
+arguments specific to the plugin's execution. Other than that the external plugin will be opaque to the Indexer--it
+will only know how to send and receive data and ensure that a block was acted on by the plugin.
+
+
+External plugin development will then require the Indexer to provide a lightweight process manager, process plugin
+templates, and a way to facilitate streaming of block data in and out of these processes (in addition to any other state
+data that will be useful for a process plugin to communicate).
+
+### End State
+
+Once this RFC is implemented, the Indexer will have a set of exporter plugins defined that can be substituted for one
+another. The default plugin will be the existing Postgresql exporter, but we will also initially support a no-op
+exporter and a stdout exporter which will drop blocks and echo them to stdout respectively.
+
+We will also be able to create exporter plugins external to the Indexer repository. In order to enable this we will have
+implemented a minimal version of the execution framework which is described in various parts above, and there will be
+a specific exporter implementation which will manage external plugin processes and provide them with data.
+
+
+### Prior art
+
+[prior-art]: #prior-art
+
+A lot of this RFC takes inspiration from the [Telegraf](https://github.com/influxdata/telegraf) metrics pipelines--a
+framework for server side metrics collection. They provide configurable plugins for metrics inputs, outputs, and
+aggregators/processors. Their plugin system also provides "shims" which use their custom process manager to enable
+external processes for use as plugins. These shims read/write serialized data in the plugin's specified data format to
+stdin/stdout in order communicate with the processes. Their system uses a polling model in order to receive data from
+external plugins, and has configurable flushing to control sending data to plugins.
+
+
+Kafka also has a plugin-like system with Kafka Connect. It allows external parties to define Conectors which export or
+import data to their Kafka pipeline. The users have to implement this as a subclass/interface which is provided by
+Kafka. Because Kafka runs in the JVM it can correspond directly with custom classes using this interface, but obviously
+it requires users to write Java code and use the Kafka library as a dependency which is much more restrictive than the
+external process approach.
+
+
+### Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+- How should process communication be handled?
+  - Telegraf uses stdin/stdout to communicate w/ processes. Kafka is mostly built around APIs, similarly things like
+docker/linkerd/containerd use HTTP endpoints to standardize communication (though they mostly use this for config data
+instead of streaming data). 
+  - I support using sockets--very standard and well known programming interface,and has useful libraries built
```

Comment:
Sockets are a good cross-platform way of communication.  You'll need some way of serializing even if it something simple like a type byte, 4 length bytes and the payload.

Reply:
I've left this in for now. I think the the socket libraries are going to be easy to integrate into a cross-process communication framework, but obviously that needs to be given more detail and thought, and it will evolve a bit as we develop. 

We can always add another RFC for this and keep this one for historical purposes, or decide that we want to update old RFCs to reflect the actual state of things in the future.